packages/zoe/src/contractFacet/zcfZygote.js

```diff
@@ -328,29 +242,66 @@ export const makeZCFZygote = (
 
   // handleOfferObject gives Zoe the ability to notify ZCF when a new seat is
   // added in offer(). ZCF responds with the exitObj and offerResult.
-  /** @type {HandleOfferObj} */
-  const handleOfferObj = Far('handleOfferObj', {
-    handleOffer: (invitationHandle, zoeSeatAdmin, seatData) => {
-      const zcfSeat = makeZCFSeat(zoeSeatAdmin, seatData);
-      const offerHandler = takeOfferHandler(invitationHandle);
-      const offerResultP = E(offerHandler)(zcfSeat, seatData.offerArgs).catch(
-        reason => {
+  const makeHandleOfferObj = vivifyKind(
+    zcfBaggage,
+    'handleOfferObj',
+    () => ({}),
+    {
+      handleOffer: (_context, invitationHandle, seatData) => {
+        const zcfSeat = makeZCFSeat(seatData);
+        // TODO: provide a details that's a better diagnostic for the
+        // ephemeral offerHandler that did not survive upgrade.
+        const offerHandler = takeOfferHandler(invitationHandle);
+        const offerResultP =
+          typeof offerHandler === 'function'
+            ? E(offerHandler)(zcfSeat, seatData.offerArgs)
+            : E(offerHandler).handle(zcfSeat, seatData.offerArgs);
+
+        const offerResultPromise = offerResultP.catch(reason => {
           if (reason === undefined) {
             const newErr = new Error(
               `If an offerHandler throws, it must provide a reason of type Error, but the reason was undefined. Please fix the contract code to specify a reason for throwing.`,
             );
             throw zcfSeat.fail(newErr);
           }
           throw zcfSeat.fail(reason);
-        },
-      );
-      const exitObj = makeExitObj(seatData.proposal, zcfSeat);
-      /** @type {HandleOfferResult} */
-      return harden({ offerResultP, exitObj });
+        });
+        const exitObj = makeExitObj(seatData.proposal, zcfSeat);
+        /** @type {HandleOfferResult} */
+        return harden({ offerResultPromise, exitObj });
+      },
     },
-  });
+  );
+  const handleOfferObj = makeHandleOfferObj();
+
+  const evaluateContract = () => {
+    let bundle;
+    if (passStyleOf(contractBundleCap) === 'remotable') {
+      const bundleCap = contractBundleCap;
+      // @ts-expect-error vatPowers is not typed correctly: https://github.com/Agoric/agoric-sdk/issues/3239
+      bundle = powers.D(bundleCap).getBundle();
+    } else {
+      bundle = contractBundleCap;
+    }
+    return evalContractBundle(bundle);
+  };
+  // evaluate the contract (either the first version, or an upgrade)
+  const { start, buildRootObject, vivify } = await evaluateContract();
+
+  if (start === undefined && vivify === undefined) {
+    assert(
+      buildRootObject === undefined,
+      'Did you provide a vat bundle instead of a contract bundle?',
+    );
+    assert.fail('unrecognized contract exports');
+  }
+  assert(!start || !vivify, 'do not provide both start and vivify');
```

Comment:
this doesn't say that one is required.
```suggestion
  assert(!start || !vivify, `contract must provide 'start' XOR 'vivify'`);
```

Reply:
Please don't use backquotes for the details of an assert without an `X`. IIUC, you switched to backquotes because the string we'd like contains single quotes. If that's the reason, then use double quotes rather than backquotes. Or use double quotes inside and keep single quotes outside. 

How about

```suggestion
  assert(!start || !vivify, 'contract must provide exactly one of "start" or "vivify"');
```

Reply:
done